src/shred_network/shred_receiver.zig

```diff
@@ -41,68 +42,102 @@ pub const ShredReceiver = struct {
 
+    /// A shared instance of an event to support waiting on multiple Channels.
+    const ReceiverSignal = struct {
+        event: std.Thread.ResetEvent = .{},
+        hook: Channel(Packet).SendHook = .{ .after_send = afterSend },
+
+        fn afterSend(hook: *Channel(Packet).SendHook, _: *Channel(Packet)) void {
+            const signal: *ReceiverSignal = @alignCast(@fieldParentPtr("hook", hook));
+            signal.event.set();
+        }
+
+        fn wait(signal: *ReceiverSignal, exit: ExitCondition) error{Exit}!void {
+            while (true) {
+                signal.event.timedWait(1 * std.time.ns_per_s) catch {};
+                if (exit.shouldExit()) return error.Exit;
+                if (signal.event.isSet()) return signal.event.reset();
+            }
+        }
+    };
+
     /// Run threads to listen/send over socket and handle all incoming packets.
     /// Returns when exit is set to true.
     pub fn run(self: *Self) !void {
         defer self.logger.err().log("exiting shred receiver");
         errdefer self.logger.err().log("error in shred receiver");
 
-        var response_sender = try SocketThread
-            .initSender(self.allocator, self.logger.unscoped(), self.repair_socket, self.exit);
-        defer response_sender.deinit(self.allocator);
-        var repair_receiver = try SocketThread
-            .initReceiver(self.allocator, self.logger.unscoped(), self.repair_socket, self.exit);
-        defer repair_receiver.deinit(self.allocator);
+        var receive_signal = ReceiverSignal{};
+        const exit = ExitCondition{ .unordered = self.exit };
+
+        // Cretae pipe from response_sender -> repair_socket
+        const response_sender = try Channel(Packet).create(self.allocator);
+        defer response_sender.destroy();
+
+        const response_sender_pipe = try SocketPipe.init(
+            self.allocator,
+            .sender,
+            self.logger.unscoped(),
+            self.repair_socket,
+            response_sender,
+            exit,
+        );
+        defer response_sender_pipe.deinit(self.allocator);
+
+        // Create pipe from repair_socket -> response_receiver.
+        const response_receiver = try Channel(Packet).create(self.allocator);
+        response_receiver.send_hook = &receive_signal.hook;
+        defer response_receiver.destroy();
 
-        var turbine_receivers: [NUM_TVU_RECEIVERS]SocketThread = undefined;
-        for (0..NUM_TVU_RECEIVERS) |i| {
-            turbine_receivers[i] = try SocketThread.initReceiver(
+        const response_receiver_pipe = try SocketPipe.init(
+            self.allocator,
+            .receiver,
+            self.logger.unscoped(),
+            self.repair_socket,
+            response_receiver,
+            exit,
+        );
+        defer response_receiver_pipe.deinit(self.allocator);
+
+        // Create N pipes from turbine_socket -> turbine_receiver.
+        const turbine_receiver = try Channel(Packet).create(self.allocator);
```

Comment:
What do you think about the contention from sending to the same channel from multiple threads? Typical network conditions are about 5000 packets per second. Firedancer aims to support 1 million.

Reply:
Wrote a quick [channel benchmark](<https://gist.github.com/kprotty/8e4f57e21e80992655e26d5823ee4266>) which tests throughput under max contention (constantly sending). M4 with 12 senders is ~70mil/sec (~20mil/sec with receiver). 5900x with 24 senders is ~30mil/sec (~15mi/sec with receiver).

The channel parts at least should hold up. I think the bigger bottlenecks would be either IO or thread parking/unparking on ResetEvent.

Reply:
we should keep this in sig as a load tester